qp_qiime2/qiime2.py

```diff
@@ -59,3 +64,93 @@ def rarefy(qclient, job_id, parameters, out_dir):
     ainfo = [ArtifactInfo('o-table', 'BIOM', [(rarefied_fp, 'biom')])]
 
     return True, ainfo, ""
+
+
+def beta_diversity(qclient, job_id, parameters, out_dir):
+    """generate beta diversity calculations
+
+    Parameters
+    ----------
+    qclient : qiita_client.QiitaClient
+        The Qiita server client
+    job_id : str
+        The job id
+    parameters : dict
+        The parameter values to rarefy
+    out_dir : str
+        The path to the job's output directory
+
+    Returns
+    -------
+    boolean, list, str
+        The results of the job
+    """
+    out_dir = join(out_dir, 'beta_diversity')
+    if not exists(out_dir):
+        mkdir(out_dir)
+
+    qclient.update_job_step(job_id, "Step 1 of 4: Collecting information")
+    artifact_id = parameters['i-table']
+    metric = parameters['p-metric']
+    tree = parameters['i-tree']
+    artifact_info = qclient.get("/qiita_db/artifacts/%s/" % artifact_id)
+    biom_fpi = artifact_info['files']['biom'][0]
+    biom_qza = join(out_dir, 'q2-biom.qza')
+
+    qclient.update_job_step(
+        job_id, "Step 2 of 4: Converting Qiita artifacts to Q2 artifact")
+    # converting biom
+    cmd = ('qiime tools import --input-path %s --output-path %s '
+           '--type "FeatureTable[Frequency]' % (biom_fpi, biom_qza))
+    b = load_table(biom_fpi)
+    counts = list(map(sum, b.iter_data()))
+    if min(counts) == max(counts):
+        cmd += " % Properties(['uniform-sampling'])\""
+    else:
+        cmd += '"'
+    std_out, std_err, return_value = system_call(cmd)
+    if return_value != 0:
+        error_msg = ("Error converting biom:\nStd out: %s\nStd err: %s"
+                     % (std_out, std_err))
+        return False, None, error_msg
+    # converting tree
+    if tree is not None:
+        qza_tree = join(out_dir, 'tree.qza')
+        cmd = ('qiime tools import --input-path %s --type Phylogeny[Rooted] '
+               '--output-path %s' % (tree, qza_tree))
+        tree = qza_tree
+        std_out, std_err, return_value = system_call(cmd)
+        if return_value != 0:
+            error_msg = ("Error converting biom:\nStd out: %s\nStd err: %s"
+                         % (std_out, std_err))
+            return False, None, error_msg
+
+    qclient.update_job_step(
+        job_id, "Step 3 of 4: Calculating beta diversity: %s" % (metric))
+    dtx_fp = join(out_dir, '%s.qza' % metric)
+    if tree is not None and metric in STATE_UNIFRAC_METRICS:
```

Comment:
If the metric is in STATE_UNIFRAC_METRICS, but the tree is None this if will evaluate to false and go by default to non-phylogenetic metric, which is incorrect. I think this `if-else` block is missing an `else if tree is None and metric in STATE_UNIFRAC_METRICS` that returns an error.